File: augment_and_convert_csv.js
import fs from 'fs';
import path from 'path';

function augmentData(rows, targetRows = 200) {
    if (rows.length >= targetRows) return rows.slice(0, targetRows);

    const additionalNeeded = targetRows - rows.length;
    const newRows = [];

    for (let i = 0; i < additionalNeeded; i++) {
        const sourceRow = rows[Math.floor(Math.random() * rows.length)];
        const prefix = i % 2 === 0 ? 'Senior' : 'Junior';

        newRows.push({
            OkupasiID: `PON-AUG-${String(i + 1).padStart(3, '0')}`,
            Area_Fungsi: sourceRow.Area_Fungsi,
            Okupasi: `${prefix} ${sourceRow.Okupasi}`,
            Unit_Kompetensi: sourceRow.Unit_Kompetensi,
            Kuk_Keywords: sourceRow.Kuk_Keywords
        });
    }

    return [...rows, ...newRows];
}

function parseCSV(content, separator = ';') {
    const lines = content.trim().split('\n');
    const headers = lines[0].split(separator).map(h => h.trim().replace(/"/g, ''));
    const rows = [];

    for (let i = 1; i < lines.length; i++) {
        const values = lines[i].split(separator).map(v => v.trim().replace(/"/g, ''));
        if (values.length === headers.length) {
            const row = {};
            headers.forEach((h, idx) => { row[h] = values[idx]; });
            rows.push(row);
        }
    }
    return rows;
}

function main() {
    const csvFile = 'DTP.csv';
    const outputFile = path.join('public', 'datasets', 'DTP_Database.jsonl');

    console.log(`Reading ${csvFile}...`);
    const content = fs.readFileSync(csvFile, 'utf-8');
    const rows = parseCSV(content, ';');

    console.log(`Original rows: ${rows.length}`);
    const augmented = augmentData(rows, 200);
    console.log(`Augmented rows: ${augmented.length}`);

    // Ensure output directory exists
    fs.mkdirSync(path.dirname(outputFile), { recursive: true });

    // Write JSONL
    const jsonlContent = augmented.map(r => JSON.stringify(r)).join('\n') + '\n';
    fs.writeFileSync(outputFile, jsonlContent, 'utf-8');

    console.log(`Saved to ${outputFile}`);
    console.log('Done!');
}

main();

File: augment_and_convert_csv.py
import pandas as pd
import json
import random
import os

def augment_data(df, target_rows=200):
    current_rows = len(df)
    if current_rows >= target_rows:
        return df.head(target_rows)
    
    additional_rows_needed = target_rows - current_rows
    new_rows = []
    
    # Extract unique values to mix and match
    areas = df['Area_Fungsi'].unique().tolist()
    
    for i in range(additional_rows_needed):
        # Pick a random existing row as a template
        source_row = df.iloc[random.randint(0, current_rows - 1)]
        
        # Determine Area_Fungsi
        area = source_row['Area_Fungsi']
        
        # Generate a new Okupasi name by slightly modifying an existing one
        original_okupasi = source_row['Okupasi']
        new_okupasi = f"Senior {original_okupasi}" if i % 2 == 0 else f"Junior {original_okupasi}"
        
        # Create a new unique ID
        new_id = f"PON-AUG-{i+1:03d}"
        
        new_rows.append({
            'OkupasiID': new_id,
            'Area_Fungsi': area,
            'Okupasi': new_okupasi,
            'Unit_Kompetensi': source_row['Unit_Kompetensi'],
            'Kuk_Keywords': source_row['Kuk_Keywords']
        })
    
    augmented_df = pd.concat([df, pd.DataFrame(new_rows)], ignore_index=True)
    return augmented_df

def main():
    csv_file = 'DTP.csv'
    output_file = 'public/datasets/DTP_Database.jsonl'
    
    print(f"Reading {csv_file}...")
    # The file uses semicolon as separator based on view_file output
    df = pd.read_csv(csv_file, sep=';')
    
    print(f"Original rows: {len(df)}")
    df_augmented = augment_data(df, target_rows=200)
    print(f"Augmented rows: {len(df_augmented)}")
    
    # Ensure directory exists
    os.makedirs(os.path.dirname(output_file), exist_ok=True)
    
    print(f"Saving to {output_file}...")
    with open(output_file, 'w', encoding='utf-8') as f:
        for record in df_augmented.to_dict('records'):
            f.write(json.dumps(record) + '\n')
            
    print("Done!")

if __name__ == "__main__":
    main()
